stackit/internal/services/iaas/securitygrouprule/resource.go

```diff
@@ -0,0 +1,658 @@
+func (r *securityGroupRuleResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	idParts := strings.Split(req.ID, core.Separator)
+
+	if len(idParts) != 3 || idParts[0] == "" || idParts[1] == "" || idParts[2] == "" {
+		core.LogAndAddError(ctx, &resp.Diagnostics,
+			"Error importing security group rule",
+			fmt.Sprintf("Expected import identifier with format: [project_id],[security_group_id],[security_group_rule_id]  Got: %q", req.ID),
+		)
+		return
+	}
+
+	projectId := idParts[0]
+	securityGroupId := idParts[1]
+	securityGroupRuleId := idParts[2]
+	ctx = tflog.SetField(ctx, "project_id", projectId)
+	ctx = tflog.SetField(ctx, "security_group_id", securityGroupId)
+	ctx = tflog.SetField(ctx, "security_group_rule_id", securityGroupRuleId)
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("project_id"), projectId)...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("security_group_id"), securityGroupId)...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("security_group_rule_id"), securityGroupRuleId)...)
+	tflog.Info(ctx, "security group rule state imported")
+}
+
+func mapFields(securityGroupRuleResp *iaasalpha.SecurityGroupRule, model *Model) error {
+	if securityGroupRuleResp == nil {
+		return fmt.Errorf("response input is nil")
+	}
+	if model == nil {
+		return fmt.Errorf("model input is nil")
+	}
+
+	var securityGroupRuleId string
+	if model.SecurityGroupRuleId.ValueString() != "" {
+		securityGroupRuleId = model.SecurityGroupRuleId.ValueString()
+	} else if securityGroupRuleResp.Id != nil {
+		securityGroupRuleId = *securityGroupRuleResp.Id
+	} else {
+		return fmt.Errorf("security group rule id not present")
+	}
+
+	idParts := []string{
+		model.ProjectId.ValueString(),
+		model.SecurityGroupId.ValueString(),
+		securityGroupRuleId,
+	}
+	model.Id = types.StringValue(
+		strings.Join(idParts, core.Separator),
+	)
+
+	model.SecurityGroupRuleId = types.StringValue(securityGroupRuleId)
+	model.Direction = types.StringPointerValue(securityGroupRuleResp.Direction)
+	model.Description = types.StringPointerValue(securityGroupRuleResp.Description)
+	model.EtherType = types.StringPointerValue(securityGroupRuleResp.Ethertype)
+	model.IpRange = types.StringPointerValue(securityGroupRuleResp.IpRange)
+	model.RemoteSecurityGroupId = types.StringPointerValue(securityGroupRuleResp.RemoteSecurityGroupId)
+
+	err := mapIcmpParameters(securityGroupRuleResp, model)
+	if err != nil {
+		return fmt.Errorf("map icmp_parameters: %w", err)
+	}
+	err = mapPortRange(securityGroupRuleResp, model)
+	if err != nil {
+		return fmt.Errorf("map port_range: %w", err)
+	}
+	err = mapProtocol(securityGroupRuleResp, model)
+	if err != nil {
+		return fmt.Errorf("map protocol: %w", err)
+	}
+
+	return nil
+}
+
+func mapIcmpParameters(securityGroupRuleResp *iaasalpha.SecurityGroupRule, m *Model) error {
+	if securityGroupRuleResp.IcmpParameters == nil {
```

Comment:
Did you test the case where the user sets the object fields, e.g. `icmp_parameters` to an empty object? In this case they would be returned by the API as nil, but we should map them to the empty object otherwise there will be value conversion error. 
 
 This doesn't apply when the object field has all fields inside as `Required` because it is not allowed to set the empty object in the configuration, but please double check for all of these object fields
 
 See how we are doing it for the labels field for example

Reply:
Added empty field checks for the optional fields. Also extended unit tests